src/aspire/utils/bot_align.py

```diff
@@ -0,0 +1,181 @@
+
+    # Lookup our params
+    loss_params = LOSS_TYPES[loss_type]
+
+    # Downsample Volumes
+    vol_ref_ds = vol_ref.downsample(downsampling_level)
+    vol_given_ds = vol_given.downsample(downsampling_level)
+
+    if loss_type == "wemd":
+        embed_0 = wemd_embed(
+            vol_ref_ds.asnumpy()[0], loss_params["wavelet"], loss_params["level"]
+        )
+
+    def loss(R):
+        """
+        Alignment loss function.
+        """
+        v = vol_given_ds.rotate(Rotation(R)).asnumpy()[0]
+        if loss_type == "eu":
+            return norm(v - vol_ref_ds)
+        elif loss_type == "wemd":
+            embed_v = wemd_embed(v, loss_params["wavelet"], loss_params["level"])
+            return norm(embed_v - embed_0, ord=1)
+
+    def cf(x1, x2):
+        """
+        The squared exponential covariance function over SO(3).
+        """
+        d = norm(x1 - x2) / loss_params["lengthscale"]
+        return np.exp(-(d**2) / 2, dtype=dtype)
+
+    def cf_grad(x1, x2):
+        """
+        Gradient of the squared exponential covariance function.
+        """
+        return cf(x1, x2) * (x2 - x1) / loss_params["lengthscale"] ** 2
+
+    R = np.zeros((max_iters, 3, 3), dtype=dtype)
+    R[0] = np.eye(3, dtype=dtype)
+    manifold = pymanopt.manifolds.SpecialOrthogonalGroup(3)
+
+    C = np.zeros((max_iters, max_iters), dtype=dtype)
+    C[0, 0] = cf(R[0], R[0])
+
+    y = np.zeros(max_iters, dtype=dtype)
+    y[0] = loss(R[0])
+
+    for t in range(1, max_iters):
+        q = np.linalg.solve(C[:t, :t] + tau * np.eye(t, dtype=dtype), y[:t])
+
+        @pymanopt.function.numpy(manifold)
+        def cost(new, t=t, q=q):
+            """
+            Loss function for surrogate problems.
+            """
+            kx = np.array([cf(new.astype(dtype, copy=False), R[j]) for j in range(t)])
+            mu = kx @ q
+            return mu
+
+        @pymanopt.function.numpy(manifold)
+        def eu_grad(new, t=t, q=q):
+            """
+            Gradient of the loss function for surrogate problems
+            """
+            kx_grad = np.array(
+                [cf_grad(new.astype(dtype, copy=False), R[j]) for j in range(t)]
+            )
+            grad = np.einsum("ijk,i", kx_grad, q)
+            return grad
+
+        problem = pymanopt.Problem(manifold, cost, euclidean_gradient=eu_grad)
+        optimizer = pymanopt.optimizers.SteepestDescent(
+            max_iterations=man_max_iter,
+            min_gradient_norm=man_min_grad,
+            min_step_size=man_min_sz,
+            verbosity=verbosity,
+        )
+        result = optimizer.run(problem)
+        R_new = result.point.astype(dtype, copy=False)
+
+        y[t] = loss(R_new)
+        R[t] = R_new
+        k = np.array([cf(R_new, R[i]) for i in range(t + 1)])
+        C[t, 0:t] = k[0:t]
+        C[0 : t + 1, t] = k
+
+    idx = y.argmin(0)
+    R_init = R[idx]
+
+    if downsampling_level > 32:
```

Comment:
Again with the hardcoded constants. This feels like it should be set somewhere.

Reply:
@RuiyiYang , is this value critical?

Have you actually explored the algorithm for other values here?

Reply:
I have added an argument for this. We can add any useful remarks to the docstring/comments.